Moved LUTs to be external to the module.

aehart opened the pull request:
This PR is meant to start a discussion about moving the lookup tables out of certain processing modules in order to greatly reduce the time required to run C-synthesis and export IP cores. There are 579 TrackletEngines in the current project, so I started with it, since we have a lot to gain here in terms of reducing this time.

The `pttable`, `bendinnertable`, and `bendoutertable` are now just memory interfaces on the TE IP core. The actual tables will exist as ROMs in the EMP payload, which will be wired to the appropriate TE by the project generation scripts. For example, here is what the `pttable` ports look like:
| RTL ports | Dir | Bits | Protocol | Source object | C type |
| --- | --- | --- | --- | --- | --- |
| pttable_V_address0 | out | 5 | ap_memory | pttable_V | array |
| pttable_V_ce0 | out | 1 | ap_memory | pttable_V | array |
| pttable_V_q0 | in | 1 | ap_memory | pttable_V | array |

I think we will still need a few different IP cores for the TE because the bit widths vary for different types of VM stubs (PS, 2S, disk, etc.). But going from 579 IP cores to three or four will still save us a lot of CPU-hours.

The TE still passes C-simulation, is still pipelined with II=1, and still passes C/RTL cosimulation. The resource utilization and timing estimates have also barely changed (the previously-internal lookup tables are quite small and were implemented as distributed ROMs):

### HLS estimates (VU7P @ 250 MHz)
| | Internal LUTs | External LUTs |
| --- | --- | --- |
| BRAM_18K | 0 | 0 |
| DPS48E | 0 | 0 |
| FF | 891 | 898 |
| LUT | 5554 | 5545 |
| URAM | 0 | 0 |
| Clock period | 6.035 ns | 6.035 ns |

### Post-implementation estimates (VU7P @ 250 MHz)
| | Internal LUTs | External LUTs |
| --- | --- | --- |
| CLB | 170 | 158 |
| LUT | 761 | 726 |
| FF | 613 | 613 |
| DSP | 0 | 0 |
| BRAM | 0 | 0 |
| SRL | 0 | 7 |
| URAM | 0 | 0 |
| Clock period | 3.158 ns | 3.160 ns |

pwittich commented:
very nice. can you test this in the context of the TE-TC chain and see what happens?

aehart commented:
Yeah, that's the next step. I'll try plugging this version of the TE into the standalone TE-TC chain and see if it works and how it compares to the current version.

pwittich commented:
do you want this PR to be reviewed before or after this step?

aehart commented:
Let's wait until after I can do this test. I should be able to do it tomorrow, and I don't see any reason to be too hasty here.

aehart commented:
I finally finished testing this new version of the TE by plugging it into the standalone TE-TC chain that we developed previously. So now, instead of 13 different IP cores, there are 13 instances of a single IP core, wired to 13 different sets of lookup tables. And I implemented the lookup tables as distributed ROMs in order to have an apples-to-apples comparison with the previous version of the TE.

The behavioural simulation continues to show that the output matches the expectation, and the resource utilization after implementation for a VU7P is very similar to what we had before:
<table>
  <tr>
    <th rowspan="2"></th>
    <th rowspan="2">Internal LUTs</th>
    <th colspan="5">External LUTs</th>
  </tr>
  <tr>
    <th>TE</th>
    <th>ptcut table</th>
    <th>stubptinnercut table</th>
    <th>stubptoutercut table</th>
    <th>Total</th>
  </tr>
<tr>
  <td>CLB LUTs</td>        <td>611</td>  <td>585</td>  <td>1</td>  <td>4</td>  <td>4</td>  <td>594</td>
</tr>
<tr>
  <td>CLB Registers</td>   <td>586</td>  <td>586</td>  <td>2</td>  <td>2</td>  <td>2</td>  <td>592</td>
</tr>
<tr>
  <td>CARRY8</td>          <td>7</td>    <td>7</td>    <td>0</td>  <td>0</td>  <td>0</td>  <td>7</td>
</tr>
<tr>
  <td>F7 Muxes</td>        <td>5</td>    <td>1</td>    <td>0</td>  <td>2</td>  <td>2</td>  <td>5</td>
</tr>
<tr>
  <td>F8 Muxes</td>        <td>0</td>    <td>0</td>    <td>0</td>  <td>1</td>  <td>1</td>  <td>2</td>
</tr>
<tr>
  <td>CLB</td>             <td>135</td>  <td>134</td>  <td>1</td>  <td>2</td>  <td>1</td>  <td>138</td>
</tr>
<tr>
  <td>LUT as Logic</td>    <td>611</td>  <td>578</td>  <td>1</td>  <td>4</td>  <td>4</td>  <td>587</td>
</tr>
<tr>
  <td>LUT as Memory</td>   <td>0</td>    <td>7</td>    <td>0</td>  <td>0</td>  <td>0</td>  <td>7</td>
</tr>
<tr>
  <td>Block RAM Tile</td>  <td>0</td>    <td>0</td>    <td>0</td>  <td>0</td>  <td>0</td>  <td>0</td>
</tr>
<tr>
  <td>DPSs</td>            <td>0</td>    <td>0</td>    <td>0</td>  <td>0</td>  <td>0</td>  <td>0</td>
</tr>
</table>

I also removed some now obsolete template parameters and comments. I think this now ready to merge.

pwittich commented:
Looks good, can you do a rebase to the master?

aehart commented:
Done!
